core/secrets/secretvalue.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package secrets
+
+import (
+	"encoding/base64"
+	"io/ioutil"
+	"strings"
+
+	"github.com/juju/errors"
+)
+
+// SecretValue holds the value of a secret.
+// Instances of SecretValue are returned by a secret store
+// when a secret look up is performed. The underlying value
+// is a map of base64 encoded values represented as []byte.
+// Convenience methods exist to retrieve singular decoded string
+// and encoded base64 string values.
+type SecretValue interface {
+	// EncodedValues returns the key values of a secret as
+	// the raw base64 encoded strings.
+	// For the special case where the secret only has a
+	// single key value "data", then use BinaryValue()
+	//to get the result.
+	EncodedValues() map[string]string
+
+	// Values returns the key values of a secret as strings.
+	// For the special case where the secret only has a
+	// single key value "data", then use StringValue()
+	//to get the result.
+	Values() (map[string]string, error)
+
+	// Singular returns true if the secret value represents a
+	// single data value rather than key values.
+	Singular() bool
+
+	// EncodedValue returns the value of the secret as the raw
+	// base64 encoded string.
+	// The secret must be a singular value.
+	EncodedValue() (string, error)
+
+	// Value returns the value of the secret as a string.
+	// The secret must be a singular value.
+	Value() (string, error)
+}
+
+type secretValue struct {
+	// Data holds the key values of a secret.
+	// We use a map to hold multiple values, eg cert and key
+	// The serialised form of any string values is a
+	// base64 encoded string, representing arbitrary values.
+	data map[string][]byte
+}
+
+// NewSecretValue returns a secret using the specified map of values.
+// The map values are assumed to be already base64 encoded.
+func NewSecretValue(data map[string][]byte) SecretValue {
```

Comment:
This approach has lead us towards anti-patterns that we are hard pressed to reverse in existing code.
We should lean towards "Accept interfaces; return structs", with interfaces defined closer to usage sites.